### ISSUE TITLE
Taraxa capability refactor proposal

### ISSUE THREAD

**Issue by JakubFornadel** [Taraxa-project/taraxa-node#911]
## Is your feature request related to a problem? Please describe.

In light of our recent findings about network being stuck due to delays in processing of consensus-related packets I have a proposal. 

It is assumed that consensus is not able to proceed because the votes are not processed fast enough. That is a consequence of our current design, in which we basically process all packets in single thread and if processing of some packets take too long (might be various reasons), consensus gets stuck.

## Existing solution
We have separate thread for low level networking stuff like receiving packets, etc... Once the packet is received it is moved to the thread pool with single thread in taraxa capability, which means the networking thread is not blocked anymore. Although this solution separates networking thread from processing thread, it does not really solves our problem because all of the packets are now in single processing thread and they are blocking each other there because they are processed synchronously. If some of them take too long to process it again delays processing of votes, which ultimately leads to the network being stuck.    

### Priority queue
There could be a concurrent priority queue, in which we would save packets RLP representations together with it's priority and packet type. So once the networking thread receives new packet all it does is that it parses packet's type, assigns a priority to it and saves such data(packet RLP + type + priority) to the priority queue. 

- Consensus related packets - highest priority
- New Txs/Blocks packets - intermediate priority
- StatusPacket, Syncing packets - lowest priority 

### Real thread pool
Once we have such priority queue, there would be a threadpool with multiple threads and these threads would concurrently process packets from queue. Note that not all of them can be processed in truly concurrent way (e.g syncing packets) but such type of synchronization can be part of polling logic from priority queue...  Not all code is ready for such processing, but we can start step by step. For example StatusPacket/GetPbftBlockPacket/... can be processed concurrently with only few changes even now... 

#### Polling from queue
As the name of the queue suggest - "priority queue", we would assign different priorities to different packets. For example consensus-related packets would have the highest priority, which means it would be always polled as first even if some other packets with lower priority (e.g. syncing packets) were supposed to be processed earlier in terms of time...

There would be also other type of synchronization in polling, which would take into consideration dependencies of different packets on each other or the requirement of some specific packet to be processed synchronously(e.g. syncing)...

#### Processing of polled data 
There would be a registered handler function for each type of packet. Ultimately each current "case :" in taraxa_capability would be a different handler. Once we poll packet from queue, we know we can start to process it immediately as all the synchronization logic is done in queue polling. It serves only such packets that are not blocked by ongoing processing of some other packets...

This way we would also break that gigantic function for processing packets into smaller functions (handlers).

#### Pseudo code 
```cpp
// Networking thread
packet          = async_read()
packet_type 	 = getPacketType(packet)
packet_priority = getPacketPriority(packet_type)

priority_queue.push(packet_type, packet_priority, packet)


// Thread pool processing threads 
[packet_type, packet_rlp] = priority_queue.pop()

handler = getHandler(packet_type)
handler(packet_rlp)
```

#### Ultimate goal
We could ultimately get even rid of all those separate threads we are having at the moment (e.g. threads for validating dag blocks, transactions, etc...). Threads are in general very expensive and most of our threads are usually not used. For example even threads that validate new dag blocks are most of the time(in terms of processor time) just waiting... 

Sure we would not just refactor everything right away but to have such threadpool that could process packets as they come without the need of intermediate queues might be an interesting idea and we would use those threads in much more efficient way than we do now.

#### Pro's

- We can easily prioritize certain packets so for example consensus related packets would be processed immediately - no delays for consensus
- much more efficient threads usage (in case we start to remove some of the intermediate queues + threads that poll from them)
- real concurrent processing of packets

#### Con's

- To really take advantage of this design, we would have to adjust some parts of code so it can be truly processsed in concurrent way. Some of the changes would be trivial like for StatusPacket or  GetPbftBlockPacket for example, others might be more difficult. If we made processing of all consensus packets concurrent-friendly that should ultimately fix all problems related to the delays we are facing at the moment.
